modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py

```diff
@@ -84,27 +85,27 @@ def apply(self, func, *args, **kwargs):
         -----
         The keyword arguments are sent as a dictionary.
         """
-        call_queue = self.call_queue + [[func, args, kwargs]]
+        call_queue = self.call_queue + [Invocable(func=func, args=args, kwargs=kwargs)]
+        call_queue = [DaskWrapper.materialize_invocable(inv) for inv in call_queue]
```

Comment:
Adding this line was necessary because if we don't materialize here, `DaskWrapper.deploy` ends up getting Dask futures instead of `Invocable` objects, and those `Invocable` objects also end up having futures in its arguments.

Reply:
Thanks for figuring this out. I wonder if you could run some test workflows to benchmark whether this change has any meaningful performance regression - although I agree with your intuition that it likely won't matter.

Another really messy (and possibly confusing) way is to not compose `call_queue` with `Invocables`, and instead with the original `(func, args, kwargs)`.

That means we go from `(func, args, kwargs)` in the API layer, convert to `Invocable` when passing them around the partition layer, and unpack them when dealing with the execution layer, which seems bad. Again, all of this is to make the partition layer cleaner, at the cost of... making the partition layer complex again. What do you think?

Reply:
this should be a comment in-code somewhere for sure

Reply:
@vnlitvinov makes sense, I'll add it in.

@jeffreykennethli I'll see if I can find any performance changes. I'm not sure if it's possible to decouple `Invocable` from the call queue for Dask without making any changes in the `PandasDataframeAxisPartition` superclass, since that class constructs `Invocable`s and passes it to overridden functions; making the Dask virtual partition methods not use `Invocables` in the call queue would make reasoning about nested `Invocable`s (which take other `Invocable`s as function arguments) difficult.

Reply:
Here's a quick and dirty benchmark for ten runs of this code, where the call queue reaches depth ~1000:
```
df = pd.concat([pd.Series(np.ones(10)), pd.Series(np.ones(10))], axis=1, ignore_index=True)
for i in range(1000):
    df = pd.concat([df, pd.Series(np.ones(10))], axis=1, ignore_index=True)
repr(df)
```
This branch (`a7ce90b7`): 19.60 s
Current master (`daa50b12`): 20.16 s

This branch is 3% faster over these samples, but I don't think the setup is rigorous enough to really consider it a speedup -- at the very least, this test seems to indicate no appreciable drop in performance.

Reply:
I am not sure we want to materialize these here instead of allowing Dask to do that in a worker process. This causes inter-process communication to get materialized data in the main thread instead of a worker process where we are not blocked. Imagine a case when we have a cluster related case. The data may be transferred between nodes to get the data in the main thread.

This might be a related issue - https://github.com/dask/distributed/issues/4842. Probably, we can ping Dask guys to replay activity there.

Reply:
In that case, it seems like this is something that has to be fixed on Dask's end, since the tree-recursive structure induced by `Invocable(func, args=[Invocable(...), ...])` isn't something we can easily schedule ourselves without significant changes.

Do you have any suggestions for a workaround @YarShev? We'd like to get this PR in to make the `apply` interface uniform between block and axis partitions, and I'm not sure what a clean way to do that without materializing early would be.

Reply:
I do not think we should introduce the Invocable concept now because that hits perfromance. I suggest keeping just `func, *args, **kwargs`, which can be handled both Ray and Dask automatically, for partition.apply.